Add suport for linux and Darwin arm64/aarch64

Themimitoof opened the pull request:
Hello,

I'm trying `opkssh` on few machines I manage (linux amd64, linux arm64, openbsd amd64, darwin arm64) and I opening this pull request to add the support of ARM64 for Linux machines.

I tested with a Raspberry Pi 4 as a server, the authentication is working properly. Here's a fragment of my sshd logs:

```
Mar 26 14:17:31 pi-docker1 sshd[2665269]: Connection closed by authenticating user ubuntu [REDACTED] port 33816 [preauth]
Mar 26 14:17:34 pi-docker1 sshd[2665314]: AuthorizedKeysCommand /usr/local/bin/opkssh verify ubuntu AAAAB...
Mar 26 14:17:35 pi-docker1 sshd[2665314]: Certificate extension "openpubkey-pkt" is not supported
Mar 26 14:17:35 pi-docker1 sshd[2665314]: Certificate extension "openpubkey-pkt" is not supported
Mar 26 14:17:35 pi-docker1 sshd[2665314]: Accepted publickey for ubuntu from [REDACTED] port 39400 ssh2: ECDSA-CERT SHA256:VxVHrHW7uolYu+RzpmG8pDxCl0YxvtZ61Fxdr9pkGPA ID ***@gmail.com (serial 0) CA ECDSA SHA256:VxVHrHW7uolYu+RzpmG8pDxCl0YxvtZ61Fxdr9pkGPA
Mar 26 14:17:35 pi-docker1 sshd[2665314]: pam_unix(sshd:session): session opened for user ubuntu by (uid=0)
```

The server installation script has been updated to support amd64 and aarch64 architectures.

EthanHeilman commented on README.md:
````diff
@@ -58,7 +59,7 @@ curl -L https://github.com/openpubkey/opkssh/releases/latest/download/opkssh-osx
 ```bash
-curl -L https://github.com/openpubkey/opkssh/releases/latest/download/opkssh-linux-amd64 -o opkssh; chmod +x opkssh
+curl -L https://github.com/openpubkey/opkssh/releases/latest/download/opkssh-linux-$(uname -p) -o opkssh; chmod +x opkssh
 ```
````
On ubuntu and centos `uname -p` resolves to `x86_64` not `arm64` which breaks the install

EthanHeilman commented on scripts/install-linux.sh:
```diff
@@ -13,6 +13,15 @@ else
 
+
+# Check CPU architecture
+CPU_ARCH=$(uname -p)
```
As noted in previous comment 
uname -p resolves to x86_64 to arm64

Themimitoof replied:
Fixed with aaa78f8

EthanHeilman reviewed (changes requested):
Thank you for this PR. There are some issues with the URL.

Would it be possible to add arm/aarch64 to test matrix for the integration tests.

You could probably copy the pattern I used here  in the go code:
https://github.com/openpubkey/opkssh/blob/cf858cd9cfa954a290a0d6e3e224b59c3a34f4d9/test/integration/ssh_server/ssh_server.go#L48-L54

and here in the github-action
https://github.com/openpubkey/opkssh/blob/cf858cd9cfa954a290a0d6e3e224b59c3a34f4d9/.github/workflows/ci.yml#L32-L40

If you don't have test to add the integration test, don't worry about it. I'll open a issue to track it and add it tomorrow.

EthanHeilman commented:
@Themimitoof Let me when you plan to finish this up

Themimitoof commented:
Hi @EthanHeilman, I will finish this PR and take a look on #42 next week!

EthanHeilman commented:
@Themimitoof Is this still under development?

Themimitoof commented:
Hi @EthanHeilman!

Sorry for the delay, I made few changes:

 - I renamed the binary in `arm64` instead of `aarch64`, I guess, arm64 is more common than aarch64
 - I added a architecture check and mapping on the install script
 - I added integration tests on arm64 using the public beta of arm64 runners
 - I also added a build of MacOS on Apple Silicon as I tested on my personal machine and the client work without any issues

The CI is all green 🥳

EthanHeilman commented:
@Themimitoof Thanks so much. I going to merge the Arch linux pr in a few minutes, so you'll likely have a merge conflict, but after that you should be good to go. I'll do one last review of this today

EthanHeilman commented on README.md:
````diff
@@ -66,7 +68,7 @@ curl -L https://github.com/openpubkey/opkssh/releases/latest/download/opkssh-osx
 
 ```bash
-curl -L https://github.com/openpubkey/opkssh/releases/latest/download/opkssh-linux-amd64 -o opkssh; chmod +x opkssh
+curl -L https://github.com/openpubkey/opkssh/releases/latest/download/opkssh-linux-$(uname -p) -o opkssh; chmod +x opkssh
````
This doesn't work on amd64 (x86_64)

![image](https://github.com/user-attachments/assets/08e9600b-af4f-46ca-bdce-9a5293156ec9)

Maybe just change this to say for amd64 do X and for arm64 do X

EthanHeilman replied:
```suggestion
curl -L https://github.com/openpubkey/opkssh/releases/latest/download/opkssh-linux-amd64 -o opkssh; chmod +x opkssh

or for ARM
curl -L https://github.com/openpubkey/opkssh/releases/latest/download/opkssh-linux-arm64 -o opkssh; chmod +x opkssh
```

EthanHeilman commented:
@ajesipow 
Hitting this pacman error again which I first saw in https://github.com/openpubkey/opkssh/pull/79

ajesipow commented:
> edit 2: Does the Arch dockerfile (quay.io/archlinux/archlinux) we are using not support arm64?

It looks like so unfortunately, [on dockerhub](https://hub.docker.com/_/archlinux/) they explicitly mention only supporting `am64`. There is [this effort](https://github.com/lopsided98/archlinux-docker) supporting other architectures but may not be ideal to use a non-official image.

Themimitoof commented on .github/workflows/ci.yml:
```diff
@@ -33,10 +33,11 @@ jobs:
     strategy:
       matrix:
+        runs_on: [ubuntu-latest, ubuntu-24.04-arm]
         os: [ubuntu, centos, arch]
```
I don't think there will be breaking changes between the amd64 and arm64 version of a distribution. I suggest to disable ArchLinux on ARM64:

```suggestion
        os: [ubuntu, centos, arch]
        exclude:
          - runs_on: ubuntu-24.04-arm
             os: arch
```

Foxboron commented:
Yes, it's a good idea to treat anything that is not amd64 from Arch as a separate distribution.

EthanHeilman reviewed (approved):
Looks good!
